src/modules/mc_rate_control/MulticopterRateControl.cpp

```diff
@@ -214,7 +214,38 @@ MulticopterRateControl::Run()
 			}
 
 			// run rate controller
-			const Vector3f att_control = _rate_control.update(rates, _rates_setpoint, angular_accel, dt, _maybe_landed || _landed);
+			Vector3f att_control = _rate_control.update(rates, _rates_setpoint, angular_accel, dt, _maybe_landed || _landed);
+
+			// low pass filter the roll, pitch, and yaw actuator controls just before publishing
+			if (_param_mc_roll_cutoff.get() > 0.01f) {
```

Comment:
There is a safer function in the `AlphaFilter` class to set the cutoff frequency: https://github.com/PX4/PX4-Autopilot/blob/0afd93a9e9696be0f86dd99c7e9fa9b04066df35/src/lib/mathlib/math/filter/AlphaFilter.hpp#L76

Also, it would be better to only call that function if the sampling or cutoff frequency changed significantly (both should be stable during most flights, so updating at each loop would be a waste of CPU).

Here is an example of how to implement it: https://github.com/PX4/PX4-Autopilot/blob/0afd93a9e9696be0f86dd99c7e9fa9b04066df35/src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp#L431-L432

Reply:
Thank you for the feedback, @bresch.  I like that the current approach for setting the filter parameters uses the actual `dt` in the function which makes it robust to changes in timing.  If we are going to update the filter parameters less frequently then I think we would want some estimate of the average rate that MulticopterRateControl.cpp updates and also monitor how much that is changing.  Does that update rate exist somewhere?

Regarding safety note that the range the range for dt is checked in with this line:
`const float dt = math::constrain(((now - _last_run) * 1e-6f), 0.000125f, 0.02f);`

and the cutoff values are checked to be positive.  For example:
`_param_mc_yaw_cutoff.get() > 0.01f`

I don't like this check in setCutoffFreq which causes the parameter setting to fail suddenly if a user sets a high cutoff frequency:
`(cutoff_freq >= sample_freq / 2.f)`

With the current approach, the actual cutoff frequency of the low pass filter is just less accurate if a user sets a high cutoff frequency which I think is preferred.

Reply:
The sampling time of the rate controller is really stable as it is locked on the gyro update rate. As in https://github.com/PX4/PX4-Autopilot/blob/0afd93a9e9696be0f86dd99c7e9fa9b04066df35/src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp#L119
we can then update the parameters if the dt or the cutoff changes significantly

Reply:
Hi, @bresch .  Thank you for the feedback.  Unfortunately, the dt value is not super stable on the platform I am using now (ModalAI VOXL2).  Please see the dt value in MulticopterRateControl.cpp for several seconds plotted in the figure below.  In general, I don't think we should rely on this dt value being super stable across all platforms that run PX4.

I don't think the current approach of setting the alpha in the filter based on the current value of dt is computationally expensive (it is just a couple algebraic operations).  Also, the current approach has multiple checks to make sure alpha is a safe value.

You seem to be advocating for setting the alpha value in the filter less frequently.  We could do so by calculating the average for the last N samples of dt in MulticopterRateControl.cpp (or use some type of low pass filter on dt).  We could then update the RPY filter constants if this filtered value of dt changes significantly.  Unfortunately, we won't have data to see what this dt value looks like across all PX4 platforms to determine the logic we should use for detecting a significant change.  This type of strategy is possible to implement but it is more work without much benefit (in my view) than just simply setting the alpha based on the current dt like we are doing now.  

What do you all think, @bresch and @dagar ?

![dt_plot](https://github.com/PX4/PX4-Autopilot/assets/107884356/5fac1e26-94c5-4bc5-8e9e-7f2ad5cc7689)

Reply:
I wouldn't mind trying to eliminate that little bit of jitter on VOXL2 (ESC driver in same thread?), but even then maybe the important takeaway here is that we can't guarantee it's perfect everywhere.

A few thoughts...
 - I would propose we push more logic into AlphaFilter itself to make it easy to use, hard to misuse
    - this could be something like `update(timestamp, value)` and internally the filter parameters are updated if the dt is sufficiently different (or passing dt directly)
 - if it helps we could include a `dt` in `vehicle_angular_velocity` that's the online averaged version the gyro lowpass, notch, etc actually use
 - AlphaFilter<Vector3f> might still be viable here with the separate cutoffs per axis passed in?